Full repo manager

jimmylai opened the pull request:
## Summary
Add Full repo manager.
Also add Sphinx documents.
https://4831-200896124-gh.circle-artifacts.com/0/doc/metadata.html
![image](https://user-images.githubusercontent.com/3840867/71535747-1a190a00-28bd-11ea-9b1a-faf49d90e90d.png)



## Test Plan
Integration test:
![image](https://user-images.githubusercontent.com/3840867/71493252-1cfaf880-27f2-11ea-8dfd-59e9b81dc659.png)

Also used the test suite data and mock the IPC calls for unit tests.

jimmylai committed:
rebase onto pyre_test_suite

jimmylai committed:
add FullRepoManager

jimmylai committed:
add FullRepoManagerTest

mvismonte commented on libcst/metadata/full_repo_manager.py:
```diff
@@ -0,0 +1,107 @@
+        for provider in providers:
+            handler = self._get_cache_handler(provider)
+            if handler:
+                self._cache[provider] = handler(paths)
```
Correct me if I'm wrong, but this looks like it's making shell calls in the constructor.  IMO, it's best to do expensive processing in another method.  Constructing objects should be light

jimmylai replied:
I understand this concern, we need to run the batched pyre query before user calling.
If we want to make the constructor cheap, we can add one resolve step and cache the result.
```
m = FullRepoManager(root, paths, providers)
m.get_metadata_wrapper_for_path(path)  # it implicitly resolve the cache and store for reuse.
```

jimmylai replied:
The implicit resolve step is added as `_resolve_cache()`

mvismonte commented on libcst/metadata/full_repo_manager.py:
```diff
@@ -0,0 +1,107 @@
+        maps: Dict["ProviderT", Callable[[List[str]], Mapping[str, object]]] = {
+            TypeInferenceProvider: self._handle_pyre_cache
+        }
+        return maps.get(provider)
```
Should we make this map an argument to the constructor instead?  I notice that in your tests that this method is patched.  If we make the maps an argument, then we can avoid patching and pass in our own handler method.

jimmylai replied:
This method is not patched and it's run in the tests.
The patched method is `_handle_pyre_cache` which it requires actually query pyre daemon (which is an integration test). We don't want to build an integration test as a unit test to avoid flaky or slow test.

mvismonte replied:
Thanks for moving the expensive calls out!  This definitely feels better.  I think I wasn't as clear on the other part before.  I'm wondering if we should pass in the map you define on line 65 into the constructor as well.  This way, we can pass in a lambda function as the handler function for `TypeInferenceProvider` instead of mocking out `_handle_pyre_cache`.  In the long term, this feels to me so that we don't bloat this class with handler function implementations, but you've thought about this API design a bit more than I have.

jimmylai replied:
The current implementation keeps `TypeInferenceProvider` shell out logic in FullRepoManager. The other way is to keep the logic in `TypeInferenceProvider` to keep related logic together.

mvismonte reviewed (changes requested):
Looks good overall.  Just a few questions about the API design

jimmylai committed:
add documents for TypeInferenceProvider and FullRepoManager

mvismonte commented on libcst/metadata/full_repo_manager.py:
```diff
@@ -0,0 +1,154 @@
+            when timeout.
+        """
+        self.root_path: Path = Path(repo_root_dir)
+        self._cache: Dict["ProviderT", Mapping[str, object]] = {}
```
I  guess there's no way of specifying that the type of the value of the map is whatever is provided by `ProviderT`?

jimmylai replied:
We didn't think of  an easy way to do that. One idea we had to annotate the type of cache for each provider is to add a second generic type variable in provider class definition for the cache type. This approach requires inspect the second generic type at runtime which make it complex.

mvismonte reviewed (approved):
Comments inline, but this looks good overall

jimmylai committed:
move type inference related logic to TypeInferenceProvider

jimmylai commented:
Addressed all comments and moved type inference related logic to `TypeInferenceProvider`.
From now on each metadata provider requires cache needs to implement `gen_cache()` and `FullRepoManager` calls `gen_cache` to batch compute required cache.
